boskos/boskos.go

```diff
@@ -0,0 +1,383 @@
+
+func main() {
+	flag.Parse()
+	logrus.SetFormatter(&logrus.JSONFormatter{})
+
+	r, err := NewRanch(*configPath, *storage)
+	if err != nil {
+		logrus.WithError(err).Fatal("Fail to create my ranch!")
+	}
+
+	http.Handle("/", handleDefault())
+	http.Handle("/start", handleStart(r))
+	http.Handle("/done", handleDone(r))
+	http.Handle("/reset", handleReset(r))
+	http.Handle("/update", handleUpdate(r))
+	http.Handle("/metric", handleMetric(r))
+
+	go func() {
+		for range time.Tick(time.Minute) {
+			r.logStatus()
+			r.saveState()
+		}
+	}()
+
+	logrus.Info("Start Service")
+	logrus.WithError(http.ListenAndServe(":8080", nil)).Fatal("ListenAndServe returned.")
+}
+
+func handleDefault() http.HandlerFunc {
+	return func(res http.ResponseWriter, req *http.Request) {
+		logrus.WithField("handler", "handleDefault").Infof("From %v", req.RemoteAddr)
+	}
+}
+
+func handleStart(r *Ranch) http.HandlerFunc {
+	return func(res http.ResponseWriter, req *http.Request) {
+		logrus.WithField("handler", "handleStart").Infof("From %v", req.RemoteAddr)
+
+		if req.Method != "POST" {
+			logrus.Warning("[BadRequest]method %v, expect POST", req.Method)
+			http.Error(res, "/start only accept POST method", http.StatusBadRequest)
+			return
+		}
+
+		rtype := req.URL.Query().Get("type")
+		state := req.URL.Query().Get("state")
+		owner := req.URL.Query().Get("owner")
+		if rtype == "" || state == "" || owner == "" {
+			logrus.Warning("[BadRequest]type, state and owner must present in the request")
+			http.Error(res, "Type, state and owner must present in the request", http.StatusBadRequest)
+			return
+		}
+
+		logrus.Infof("Request for a %v %v from %v", state, rtype, owner)
+
+		resource := r.Request(rtype, state, owner)
+
+		if resource != nil {
+			resJSON, err := json.Marshal(resource)
+			if err != nil {
+				logrus.WithError(err).Errorf("json.Marshal failed : %v", resource)
+				http.Error(res, err.Error(), http.StatusInternalServerError)
+				return
+			}
+			logrus.Infof("Resource leased: %v", string(resJSON))
+			fmt.Fprint(res, string(resJSON))
+		} else {
+			logrus.Infof("No available resource")
+			http.Error(res, "No resource", http.StatusInternalServerError)
+		}
+	}
+}
+
+func handleDone(r *Ranch) http.HandlerFunc {
+	return func(res http.ResponseWriter, req *http.Request) {
+		logrus.WithField("handler", "handleDone").Infof("From %v", req.RemoteAddr)
+
+		if req.Method != "POST" {
+			logrus.Warning("[BadRequest]method %v, expect POST", req.Method)
+			http.Error(res, "/done only accept POST method", http.StatusBadRequest)
+			return
+		}
+
+		name := req.URL.Query().Get("name")
+		state := req.URL.Query().Get("state")
+		if name == "" || state == "" {
+			logrus.Warning("[BadRequest]name and state must present in the request")
+			http.Error(res, "Name and state must present in the request", http.StatusBadRequest)
+			return
+		}
+
+		if err := r.Done(name, state); err != nil {
```

Comment:
Included the owner. I think it's enough, user should not really care about, and store the timestamp, or they can hack around with the timestamp which will be bad.